src/Eval.hs

```diff
@@ -871,43 +865,43 @@ primitiveDefdynamic _ ctx [XObj (Sym (SymPath [] name) _) _ _, value] = do
     Right evaledBody ->
       dynamicOrMacroWith newCtx (\path -> [XObj DefDynamic Nothing Nothing, XObj (Sym path Symbol) Nothing Nothing, evaledBody]) DynamicTy name value
 primitiveDefdynamic _ ctx [notName, _] =
-  pure (evalError ctx ("`defndynamic` expected a name as first argument, but got " ++ pretty notName) (info notName))
+  pure (evalError ctx ("`defndynamic` expected a name as first argument, but got " ++ pretty notName) (xobjInfo notName))
 
 specialCommandSet :: Context -> [XObj] -> IO (Context, Either EvalError XObj)
-specialCommandSet ctx [(XObj (Sym path@(SymPath mod n) _) _ _), value] = do
-  (newCtx, result) <- evalDynamic ctx value
+specialCommandSet ctx [(XObj (Sym path@(SymPath mod n) _) _ _), val] = do
+  (newCtx, result) <- evalDynamic ctx val
   case result of
     Left err -> pure (newCtx, Left err)
     Right evald -> do
       let globalEnv = contextGlobalEnv ctx
       case contextInternalEnv ctx of
         Nothing -> setGlobal newCtx globalEnv evald
         Just env -> setInternal newCtx env evald
-    where setGlobal ctx env value =
+    where setGlobal ctx' env value =
             case lookupInEnv path env of
               Just (_, binder) -> do
-                (ctx', typedVal) <- typeCheckValueAgainstBinder ctx value binder
-                pure $ either (failure ctx) (success ctx') typedVal
+                (ctx'', typedVal) <- typeCheckValueAgainstBinder ctx' value binder
+                pure $ either (failure ctx'') (success ctx'') typedVal
                 where success c xo = (c{contextGlobalEnv = setStaticOrDynamicVar path env binder xo}, dynamicNil)
               Nothing -> pure (ctx, Right value)
-          setInternal ctx env value =
+          setInternal ctx' env value =
             case lookupInEnv path env of
               Just (_, binder) -> do
                 -- TODO: Type check local bindings.
                 -- At the moment, let bindings are not structured the same as global defs or dynamic defs.
                 -- This makes calls to the type check problematic, as we cannot work against a common binding form.
                 -- Once we better support let bindings, type check them.
-                (ctx', typedVal) <- typeCheckValueAgainstBinder ctx value binder
-                pure $ if contextPath ctx == mod
-                         then either (failure ctx) (success ctx') typedVal
-                         else (ctx', dynamicNil)
+                (ctx'', typedVal) <- typeCheckValueAgainstBinder ctx' value binder
+                pure $ if contextPath ctx'' == mod
+                         then either (failure ctx'') (success ctx'') typedVal
```

Comment:
I think we should use ctx'' for this one. If it fails here it's because the value and binder didn't type check, but the type information will only be present in ctx" since typeCheckValueAgainstBinder annotates the types.